walkeeper/src/callmemaybe.rs

```diff
@@ -0,0 +1,293 @@
+        let state_inner = SubscriptionStateInner {
+            handle: None,
+            last_call_time: Instant::now(),
+            paused: false,
+        };
+
+        SubscriptionState {
+            tenantid,
+            timelineid,
+            inner: Mutex::new(state_inner),
+        }
+    }
+
+    fn pause(&self) {
+        let mut state_inner = self.inner.lock().unwrap();
+        state_inner.paused = true;
+
+        if let Some(handle) = state_inner.handle.take() {
+            handle.abort();
+        }
+    }
+
+    fn resume(&self) {
+        let mut state_inner = self.inner.lock().unwrap();
+        state_inner.paused = false;
+    }
+
+    fn call(
+        &self,
+        recall_period: Duration,
+        pageserver_addr: String,
+        listen_pg_addr: String,
+        password: Option<String>,
+    ) {
+        let mut state_inner = self.inner.lock().unwrap();
+
+        // Ignore call request if this subscription is paused
+        if state_inner.paused {
+            debug!(
+                "ignore call request for paused subscription
+                tenantid: {}, timelineid: {}",
+                self.tenantid, self.timelineid
+            );
+            return;
+        }
+
+        // Check if it too early to recall
+        if state_inner.handle.is_some() && state_inner.last_call_time.elapsed() < recall_period {
+            debug!(
+                "too early to recall. state_inner.last_call_time.elapsed: {:?}, recall_period: {:?}
+                tenantid: {}, timelineid: {}",
+                state_inner.last_call_time, recall_period, self.tenantid, self.timelineid
+            );
+            return;
+        }
+
+        // If previous task didn't complete in recall_period, it must be hanging,
+        // so don't wait for it forever, just abort it and try again.
+        //
+        // Most likely, the task have already successfully completed
+        // and abort() won't have any effect.
+        if let Some(handle) = state_inner.handle.take() {
+            handle.abort();
+
+            let timelineid = self.timelineid;
+            let tenantid = self.tenantid;
+            tokio::spawn(async move {
+                if let Err(err) = handle.await {
+                    if err.is_cancelled() {
+                        warn!("callback task for timelineid={} tenantid={} was cancelled before spawning a new one",
+                            timelineid, tenantid);
+                    }
+                }
+            });
+        }
+
+        let timelineid = self.timelineid;
+        let tenantid = self.tenantid;
+        state_inner.handle = Some(tokio::spawn(async move {
+            request_callback(
+                pageserver_addr,
+                listen_pg_addr,
+                password,
+                timelineid,
+                tenantid,
+            )
+            .await
+            .unwrap_or_else(|e| {
+                error!(
```

Comment:
Seems like this error is never logged, better do that.

Reply:
uh, I was wrong and missed `error!` and `bail!` -- `error!` in the previous version was ok, but in the current code result of `request_callback` is unwrapped,  it will panic on failure.